token/program-2022/src/extension/confidential_mint_burn/instruction.rs

```diff
@@ -0,0 +1,582 @@
+#[cfg(not(target_os = "solana"))]
+use crate::proof::{process_proof_location, ProofLocation};
+#[cfg(not(target_os = "solana"))]
+use solana_zk_sdk::encryption::{auth_encryption::AeCiphertext, elgamal::ElGamalPubkey};
+#[cfg(not(target_os = "solana"))]
+use solana_zk_sdk::zk_elgamal_proof_program::{
+    instruction::ProofInstruction,
+    proof_data::{
+        BatchedGroupedCiphertext3HandlesValidityProofData, BatchedRangeProofU128Data,
+        CiphertextCiphertextEqualityProofData, CiphertextCommitmentEqualityProofData,
+    },
+};
+#[cfg(feature = "serde-traits")]
+use {
+    crate::serialization::aeciphertext_fromstr,
+    serde::{Deserialize, Serialize},
+};
+use {
+    crate::{
+        check_program_account,
+        extension::confidential_transfer::DecryptableBalance,
+        instruction::{encode_instruction, TokenInstruction},
+    },
+    bytemuck::{Pod, Zeroable},
+    num_enum::{IntoPrimitive, TryFromPrimitive},
+    solana_program::{
+        instruction::{AccountMeta, Instruction},
+        program_error::ProgramError,
+        pubkey::Pubkey,
+    },
+    solana_zk_sdk::encryption::pod::{auth_encryption::PodAeCiphertext, elgamal::PodElGamalPubkey},
+    spl_pod::optional_keys::OptionalNonZeroElGamalPubkey,
+};
+
+/// Confidential Transfer extension instructions
+#[cfg_attr(feature = "serde-traits", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "serde-traits", serde(rename_all = "camelCase"))]
+#[derive(Clone, Copy, Debug, TryFromPrimitive, IntoPrimitive)]
+#[repr(u8)]
+pub enum ConfidentialMintBurnInstruction {
+    /// Initializes confidential mints and burns for a mint.
+    ///
+    /// The `ConfidentialMintBurnInstruction::InitializeMint` instruction
+    /// requires no signers and MUST be included within the same Transaction
+    /// as `TokenInstruction::InitializeMint`. Otherwise another party can
+    /// initialize the configuration.
+    ///
+    /// The instruction fails if the `TokenInstruction::InitializeMint`
+    /// instruction has already executed for the mint.
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    ///   0. `[writable]` The SPL Token mint.
+    ///
+    /// Data expected by this instruction:
+    ///   `InitializeMintData`
+    InitializeMint,
+    /// Updates mint-authority for confidential-mint-burn mint.
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    ///   * Single authority
+    ///   0. `[writable]` The SPL Token mint.
+    ///   1. `[signer]` Confidential mint authority.
+    ///
+    ///   * Multisignature authority
+    ///   0. `[writable]` The SPL Token mint.
+    ///   1. `[]` The multisig authority account owner.
+    ///   2.. `[signer]` Required M signer accounts for the SPL Token Multisig
+    ///
+    /// Data expected by this instruction:
+    ///   `UpdateAuthorityData`
+    UpdateAuthority,
```

Comment:
Sorry I missed this on my first pass, but there is actually a `SetAuthority` [instruction](https://github.com/solana-labs/solana-program-library/blob/master/token/program-2022/src/instruction.rs#L197) that can be used to update the authority for other extensions. Let's add `ConfidentialMint` in `AuthorityType` [here](https://github.com/solana-labs/solana-program-library/blob/master/token/program-2022/src/instruction.rs#L1086) and remove the `UpdateAuthority` instruction here.

Reply:
Should I make the the confidential mint authority optional as well then? From glancing over it most the other extensions have their authority as `OptionalNonZeroPubkey`

Reply:
Yes, let's make it `OptionalNonZeroPubkey`. It seems like you already made the change. Thanks!